src/euroeval/benchmark_modules/litellm.py

```diff
@@ -341,153 +428,230 @@ def generate(self, inputs: dict) -> GenerativeModelOutput:
+            )
+            generation_kwargs["response_format"] = dict(type="json_object")
+            return
+        elif isinstance(
+            error,
+            (
                 APIConnectionError,
                 Timeout,
                 ServiceUnavailableError,
                 InternalServerError,
                 SystemError,
-            ) as e:
-                logger.debug(
-                    f"Service temporarily unavailable. The error message was: {e}. "
-                    f"Retrying in 5 seconds..."
+            ),
+        ):
+            logger.debug(
+                f"Service temporarily unavailable. The error message was: {error}. "
+                f"Retrying in 5 seconds..."
+            )
+            sleep(5)
+            return
+
+        if isinstance(error, RateLimitError):
+            raise InvalidModel(
+                f"You have encountered your rate limit for model {model_id!r}. "
+                "Skipping."
+            )
+
+        if isinstance(error, AuthenticationError):
+            raise NeedsAdditionalArgument(
+                cli_argument="--api-key",
+                script_argument="api_key=<your-api-key>",
+                run_with_cli=self.benchmark_config.run_with_cli,
+            )
+
+        raise InvalidBenchmark(
+            f"Failed to generate text. The error message was: {error}"
+        )
+
+    async def _generate_async(
+        self,
+        messages: list[dict[str, t.Any]],
+        generation_kwargs: dict[str, t.Any],
+        max_retries: int,
+        max_reruns: int,
+    ) -> tuple[list[ModelResponse], list[tuple[int, Exception]]]:
+        """Generate outputs from the model asynchronously.
+
+        Args:
+            messages:
+                The messages to pass to the model.
+            generation_kwargs:
+                The generation kwargs to pass to the model.
+            max_retries:
+                The maximum number of retries to make.
+            max_reruns:
+                The maximum number of reruns to make.
+
+        Returns:
+            A tuple containing the successful responses and the failed responses.
+        """
+        success = []
+        all_failures = {}
+        to_run = list(enumerate(messages))
+        prev_fail_count = len(to_run)
+        rerun_count = 0
+
+        while to_run and rerun_count < max_reruns and prev_fail_count > 0:
+            requests = [
+                litellm.acompletion(
+                    messages=msg, max_retries=max_retries, **generation_kwargs
                 )
-                sleep(5)
-            except AuthenticationError:
-                raise NeedsAdditionalArgument(
-                    cli_argument="--api-key",
-                    script_argument="api_key=<your-api-key>",
-                    run_with_cli=self.benchmark_config.run_with_cli,
+                for _, msg in to_run
+            ]
+            responses = await asyncio.gather(*requests, return_exceptions=True)
```

Comment:
[Supposedly](https://stackoverflow.com/a/75475366/3154226), to get a progress bar during the gather operation, we can replace this `asyncio.gather` call with `tqdm.gather`, where `tqdm` is imported as

```python
from tqdm.asyncio import tqdm
```